### ISSUE TITLE
sqliteCreateFunction method doesn't work

### ISSUE THREAD

**Issue by bohwaz** [auraphp/Aura.Sql#132]
This is funny:

```
$aura = new ExtendedPdo('sqlite::memory:');
$aura->sqliteCreateFunction('strpos', 'strpos);
// Fatal error: Call to undefined method Aura\Sql\ExtendedPdo::sqliteCreateFunction()
```

The method seems to exist only after connection:

```
class PDO_Test extends PDO {
    public function __construct() {}
    public function connect($dsn)
    {
        parent::__construct($dsn);
        $this->sqliteCreateFunction('strpos', 'strpos');
    }
}

$pdo = new PDO_Test;
var_dump(method_exists($pdo, 'sqliteCreateFunction'));
// false

$pdo->connect('sqlite::memory:');
var_dump(method_exists($pdo, 'sqliteCreateFunction'));
// true
```

I guess this would be the same with sqliteCreateAggregate and sqliteCreateCollation and other DB-specific methods of PDO.

**Comment by bohwaz** [comment]
Ah, just realised that ExtendedPdo is not using parent::__construct so I can't have access to this method even after connection, I need to use ->getPdo()->sqliteCreateFunction() instead. Makes sense, but it would also make sense to be able to use PDO methods on an "ExtendedPdo" object ;)

**Comment by pmjones** [comment]
I've been using PDO almost since it was introduced, and I did not know about this behavior. You really do learn something new every day.

**Comment by dhrrgn** [comment]
Just throwing in my 2-cents here:  Why not proxy all non-defined method calls through to the underlying PDO object?  Something like this: 

``` php
    public function __call($method, $args) {
        if (! $this->isConnected() || ! method_exists($this->pdo, $method)) {
            throw new \BadMethodCallException("Method '$method' not found on ExtendedPdo");
        }
        return call_user_func_array([$this->pdo, $method], $args);
    }
```

This would still not allow @bohwaz's first example, because the `PDO` object is lazily created.  You _could_ auto-connect in the `__call` method above.  Also, you could define this methods explicitly on `ExtendedPdo`, but that is a rabbit hole I don't think is worth venturing down...once you add 1, you will have to add them all...

_May want to do `__callStatic` too, but not sure it is needed..._

**Comment by pmjones** [comment]
> Why not proxy all non-defined method calls through to the underlying PDO object?

@dhrrgn That's a neat idea; I wonder if it would work. You could skip the BadMethodCallException entirely and just let fail. Hard to put into an interface, but then, it might not matter.

If anyone's interested in writing up a solution for this, I'd be happy to review their PR.

**Comment by bohwaz** [comment]
What I would do personally is use __call() like you said, but have a whitelist of methods and if the PDO object is NULL (= not connected), store the method name + list of arguments in an array, and in connect() iterate over these. This is slightly better in a way, as you don't have to worry, as an user, if the connection has been made or not (and you usually define functions just after constructing the object, so at this stage there is no connection). Something like this in __call:

```
static $methods = ['sqliteCreateFunction', 'sqliteCreateCollation', ...];
assert(in_array(__FUNCTION__, $methods));
$this->lazy_callbacks[__FUNCTION__][] = $args;
```

and in the connect method:

```
        foreach ($this->lazy_callbacks as $method => $callbacks)
        {
            foreach ($callbacks as $args)
            {
                call_user_func_array([$this->pdo, $method], $args);
            }
        }
```

One possible caveat is if you call the method with wrong parameters you won't have an error message until connect(), instead of having the error when calling the function, which might seem weird.

**Comment by pmjones** [comment]
The addition of methods after connection occurs with pgsql as well <https://secure.php.net/manual/en/pdo.pgsqlcopyfromarray.php> (though not with mysql or sqlsrv).

I think that the example from @dhrrgn is the one to follow here. I'll put something together.

**Comment by pmjones** [comment]
Fixed by fa03dd8 .
